src/process/base/ZLibBasedFASTQProcessorBaseClass.h

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#ifndef FASTQINDEX_ZLIBHELPER_H
+#define FASTQINDEX_ZLIBHELPER_H
+
+#include "common/CommonStructsAndConstants.h"
+#include "common/ErrorAccumulator.h"
+#include "process/extract/IndexReader.h"
+#include "process/io/Source.h"
+#include "process/io/Sink.h"
+#include "process/io/PathSource.h"
+#include <cstdio>
+#include <experimental/filesystem>
+#include <zlib.h>
+
+using std::experimental::filesystem::path;
+using std::stringstream;
+
+/**
+ * Helper class for zlib stuff.
+ */
+class ZLibBasedFASTQProcessorBaseClass : public ErrorAccumulator {
+
+protected:
+
+    shared_ptr<Source> fastqFile;
+
+    shared_ptr<Source> inputIndexFile;
+
+    shared_ptr<Sink> outputIndexFile;
+
+    /**
+     * Set to true, as soon as createIndex() was started.
+     */
+    bool wasStarted = false;
```

Comment:
The states modelled here as boolean are dependent on each other, which allows for inconsistent state combinations!

A normalized way to represent this would be with a state graph, where each node represents a combination of these booleans. A design that avoids misusage is better than one which requires knowledge about what is possible and what not.